executor.py

```diff
@@ -1,7 +1,152 @@
+    """
+    An executer to encode the images using the `comparable` hashing techniques
+    """
+
+    def __init__(
+        self,
+        hash_type: str = 'perceptual',
+        hash_size: int = 8,
+        average_hash_args: Optional[Dict] = None,
+        perceptual_hash_args: Optional[Dict] = None,
+        wavelet_hash_args: Optional[Dict] = None,
+        is_embed_bool: bool = False,
+        traversal_paths: Iterable[str] = ('r',),
+        **kwargs,
+    ):
+        """
+        :param hash_type: the hashing technique used to encode the images. By default, set to `perceptual`.
+        Possible values are: `perceptual`, `average`, `difference`, and `wavelet`.
+        Implementation of the algorithm can be found here - https://github.com/JohannesBuchner/imagehash
+        :param hash_size: the size of the encoded hash value. Should not be less than `2`
+        :param average_hash_args: the arguments for the `average` hash encoding type. By default `avg_pixels`
+        set to mean. The other possible value for `avg_pixels is `median`
+        For example: `average_hash_args={'avg_pixels': 'median'}`
+        :param perceptual_hash_args: the arguments for the `perceptual` hash encoding type. This dict should contain
+        `highfreq_factor` as the only key. The default value set is 4.
+        For example: `perceptual_hash_args={'highfreq_factors': 4}`
+        :param wavelet_hash_args: the arguments for `wavelet` hash encoding type. By default `image_scale` is equal
+         to max power of 2 for an input image and the `mode` is 'haar', which is Haar wavelets. The other possible
+         value for `mode` is db4 - Daubechies wavelets.
+         :param is_embed_bool: Set to `True` to encode the images into boolean embeddings using the hashing technique.
+         By default set to `False` to encode the images as `np.uint8` embeddings values.
+         :param traversal_paths: The default traversal path on docs, e.g. ['r'], ['c']
+        """
+        super().__init__(**kwargs)
+        if not hash_type or hash_type not in HASH_TYPE:
+            raise ValueError(
+                f'Please select one of the available `hash_type`: {HASH_TYPE}'
+            )
+        self.hash_type = hash_type
+        self.hash_size = hash_size
+
+        self._average_hash_args = average_hash_args or {}
+        if 'avg_pixels' in self._average_hash_args:
+            if self._average_hash_args['avg_pixels'] == 'mean':
+                self._average_hash_args['mean'] = np.mean
+            else:
+                self._average_hash_args['mean'] = np.median
+            del self._average_hash_args['avg_pixels']
+
+        self._perceptual_hash_args = perceptual_hash_args or {}
+        self._perceptual_hash_args.setdefault('highfreq_factor', 4)
+
+        self._wavelet_hash_args = wavelet_hash_args or {}
+        self._wavelet_hash_args.setdefault('image_scale', None)
+        self._wavelet_hash_args.setdefault('mode', 'haar')
```

Comment:
I suggest combining this part so that we only offer `hash_type` and `hash_func_args`. These functions work out of box even without these default parameters, right?

Reply:
Yes, these functions work out of the box but each hashing func have different params, so if we keep one general `hash_func_args`, it might throw error for an unwanted parameter passed. @nan-wang